src/Tools/dotnet-monitor/Egress/S3/MultiPartStream.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.S3;
+using Amazon.S3.Model;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Egress.S3;
+
+internal class MultiPartUploadStream : Stream
+{
+    private readonly byte[] _buffer;
+    private int _offset;
+    private readonly string _bucketName;
+    private readonly string _objectKey;
+    private readonly string _uploadId;
+    private readonly IAmazonS3 _client;
+    private readonly List<PartETag> _parts = new();
+    public List<PartETag> Parts => _parts.ToList();
+    public bool Disposed { get; private set; }
+    private int _position;
+    private const int MinimumSize = 5 * 1024 * 1024; // the minimum size of an upload part (except for the last part)
+
+    public MultiPartUploadStream(IAmazonS3 client, string bucketName, string objectKey, string uploadId, int bufferSize)
+    {
+        bufferSize = Math.Max(bufferSize, MinimumSize); // has to be at least the minimum
+        _buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+        _offset = 0;
+        _client = client;
+        _bucketName = bucketName;
+        _objectKey = objectKey;
+        _uploadId = uploadId;
+    }
+
+    public override async Task FlushAsync(CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException("The stream was already disposed");
+        if (_offset == 0 || _offset < MinimumSize)
```

Comment:
Is it a requirement by the AWS S3 endpoints that the last part is the only part that can be smaller than the minimum size? If not, please remove `|| _offset < MinimumSize`, remove the `FinalizeAsync` method, and allow the callers to flush whenever they want.

Reply:
Yes there is the requirement that there is no part smaller than 5 MB except the last part.

Reply:
Please provide a link to the documentation with this requirement and add it to the relevant code.